synapseclient/evaluation.py

```diff
@@ -195,12 +222,33 @@ def getURI(cls, id):
         return '/evaluation/submission/%s/status' % id
 
     def __init__(self, **kwargs):
+        annotations = kwargs.get('submissionAnnotations', {})
+        # If it is synapse annotations, turn into a format
+        # that can be worked with otherwise, create
+        # synapseclient.Annotations
+        kwargs['submissionAnnotations'] = _convert_to_annotation_cls(
+            id=kwargs['id'],
+            etag=kwargs['etag'],
+            values=annotations
+        )
+
         super(SubmissionStatus, self).__init__(kwargs)
 
     def postURI(self):
         return '/evaluation/submission/%s/status' % self.id
 
     def putURI(self):
+        # If not synapse annotations, turn them into synapseclient.Annotations
+        # must have id and etag to turn into synapse annotations
+        if not is_synapse_annotations(self.submissionAnnotations):
```

Comment:
An alternative here is to fail the store function is not a synapse annotation OR not the `Annotation` class.

Reply:
I know we talked about this on Slack, but I think I didn’t realize then that the type of the submissionAnnotations would actually be different as a side effect of the putURI (I thought it was protecting against the user setting a synapse dicitonary style submissionAnnotations attribute onto the SubmissionStatus after init but before the store).

It seems like a strange side effect, i.e. calling putURI which is presumably just a getter causes the object to morph. Could we instead override the SubmissionStatus’ json() method and put the annotation translation there? That way the SubmissionStatus itself has a consistent shape and it's only the serialization when it's washed through the store method where the translation occurs.

Reply:
I think I understand this.  But the issue is that there is nothing stopping people from doing:

```
sub_status = syn.getSubmissionStatus(123455)
sub_status.submissionAnnotations = {"foo": "bar"}
```

In this scenario above, I think you also convert dictionary annotations into `synapseclient.annotations` for entities.  I will overload the `json` function, but I'm wondering if we should throw an error if people do specify non `synapseclient.Annotation` object in `submissionAnnotation` or should I add this chunk of code in to allow for the example I posted.

Reply:
Oh, I was thinking a side effect-less overload of json rather than shifting the side effects to json(). What if we did something like the below. Then the SubmissionStatus object itself never changes unexpectedly as a side effect of storing it, just the serialized dictionary that is stored is customized for the format needed by the backend.

```python
    def json(self, ensure_ascii=True):
        json_dict = self

        if not is_synapse_annotations(self.submissionAnnotations):
            json_dict = self.copy()

            annotations = _convert_to_annotation_cls(
                id=self.id, etag=self.etag,
                values=self.submissionAnnotations
            )

            # Turn into synapse annotation
            json_dict['submissionAnnotations'] = to_synapse_annotations(
                annotations
            )

        return json.dumps(json_dict, sort_keys=True, indent=2, ensure_ascii=ensure_ascii)
```

Reply:
I did do that